app/src/main/java/org/oppia/app/parser/StringToFractionParser.kt

```diff
@@ -8,14 +12,48 @@ class StringToFractionParser {
   private val wholeNumberOnlyRegex = """^-? ?(\d+)$""".toRegex()
   private val fractionOnlyRegex = """^-? ?(\d+) ?/ ?(\d+)$""".toRegex()
   private val mixedNumberRegex = """^-? ?(\d+) (\d+) ?/ ?(\d+)$""".toRegex()
+  private val invalidCharsRegex = """^[\d\s/-]+$""".toRegex()
 
-  fun getFractionFromString(text: String): Fraction {
+  /**
+   * This method helps to validate the inputText and return [FractionParsingError]
+   * This called on submit button click.
+   * @param inputText is the user input in the [FractionInputInteractionView]
+   * @return enum [FractionParsingError]
+   */
+  fun getSubmitTimeError(text: String): FractionParsingError {
+    // No need to check for real-time errors since the following logically include them.
+    val fraction = getFractionFromString(text)
+    return when {
+      fraction == null -> FractionParsingError.INVALID_FORMAT
+      fraction.denominator == 0 -> FractionParsingError.DIVISION_BY_ZERO
+      else -> FractionParsingError.VALID
+    }
+  }
+
+  /**
+   * This method helps to validate the inputText and return [FractionParsingError]
+   * This called on text change.
+   * @param inputText is the user input in the [FractionInputInteractionView]
+   * @return enum [FractionParsingError]
+   */
+  fun getRealTimeError(text: String): FractionParsingError {
+    val normalized = text.normalizeWhitespace()
+    return when {
+      !normalized.matches(invalidCharsRegex) -> FractionParsingError.INVALID_CHARS
+      normalized.startsWith("/") -> FractionParsingError.INVALID_FORMAT
+      normalized.count { it == '/' } > 1 -> FractionParsingError.INVALID_FORMAT
+      normalized.lastIndexOf('-') > 0 -> FractionParsingError.INVALID_FORMAT
+      else -> FractionParsingError.VALID
+    }
+  }
+
+  fun getFractionFromString(text: String): Fraction? {
     // Normalize whitespace to ensure that answer follows a simpler subset of possible patterns.
     val inputText: String = text.normalizeWhitespace()
     return parseMixedNumber(inputText)
       ?: parseFraction(inputText)
       ?: parseWholeNumber(inputText)
-      ?: throw IllegalArgumentException("Incorrectly formatted fraction: $text")
+      ?: null
```

Comment:
This actually should still throw, but we should factor out the parsing to a shared helper that returns null for getSubmitTimeError, e.g.:

```kotlin
fun getFractionFromString(text: String): Fraction {
  return parseFunction(text) ?: throw IllegalArgumentException("...")
}

  private fun parseFraction(text: String): Fraction? {
    // Normalize whitespace to ensure that answer follows a simpler subset of possible patterns.
    val inputText: String = text.normalizeWhitespace()
    return parseMixedNumber(inputText)
      ?: parseFraction(inputText)
      ?: parseWholeNumber(inputText)
      ?: null
}
```

This may need fixes and reformatting, but this is the idea.

Reply:
throw IllegalArgumentException("Incorrectly formatted fraction: $text") is replaced by null because in 
```
fun getSubmitTimeError(text: String): FractionParsingError {
    // No need to check for real-time errors since the following logically include them.
    val fraction = getFractionFromString(text)
    return when {
      fraction == null -> FractionParsingError.INVALID_FORMAT
      fraction.denominator == 0 -> FractionParsingError.DIVISION_BY_ZERO
      else -> FractionParsingError.VALID
    }
  } 
``` 
Needs it to be null as per your above getSubmitTimeError()... this suggestion again will revoke previous issue of crash.

Reply:
More than that in our code 
```
 private fun parseFraction(inputText: String): Fraction? {
    val fractionOnlyMatch = fractionOnlyRegex.matchEntire(inputText) ?: return null
    val (_, numeratorText, denominatorText) = fractionOnlyMatch.groupValues
    // Fraction-only numbers imply no whole number.
    return Fraction.newBuilder()
      .setIsNegative(isInputNegative(inputText))
      .setNumerator(numeratorText.toInt())
      .setDenominator(denominatorText.toInt())
      .build()
  }
```
parseFraction() is for parsing of fraction.

Reply:
Sorry I don’t understand. Why can’t we use the two methods I suggested above to ensure the actual parsing throws an error, and the helper is used both for parsing and validation? Using the same code for both was the intent behind this approach for validation (to ensure consistency between parsing and validation).

Reply:
Done. PTAL